### PR TITLE
Add and update icons and Update icon automation (debug mode)

### PR DESCRIPTION
**Pull request by zeusongit**
### Purpose

The PR adds and updates node icons shared internally.
Total 600 icons are updated and 312 icons are added.

The size difference between old and new icons is ~9MB as we are using 400dpi icons now.

Before:
Screenshot 2024-02-14 at 7 09 49 PM

After:
Screenshot 2024-02-14 at 6 58 53 PM

Before:
Screenshot 2024-02-14 at 7 10 40 PM

After:
Screenshot 2024-02-14 at 7 10 43 PM

Better resolution:
Picture1

The new automation is under DebugMode : UpdateNodeIcons
It requires the path to new icons and will do everything else, given that:
- The icon names are in the correct format: Named after Node's FullName or IconName if it has overloads,  and suffixed by Large or Small (png)

There is a review step that will scan the directory for icons, and load them with old icons for easy comparison, logs are generated and stored as well for troubleshooting.
The second step is to update icons after the result from first review step seems satisfactory.
A backup file is also generated for the updated resx files, just in case.

DynamoSandbox_hY8x82Ws7N

DynamoSandbox_Tm0GaWUX5x

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB

### Release Notes

- Icon Refresh

### Reviewers

@DynamoDS/dynamo

### CODE REVIEW

**Comment by QilongTang** [comment]
Nice automation, just curious did you use the 100 subfolder or the 400 subfolder?

**Comment by zeusongit** [comment]
> Nice automation, just curious did you use the 100 subfolder or the 400 subfolder?

400 for all

**Comment by QilongTang** [comment]
> > Nice automation, just curious did you use the 100 subfolder or the 400 subfolder?
> 
> 400 for all

@Jingyi-Wen I think we were talking about integrating 100 ones but curious if 400 ones would bring any benefit? e.g. would  resolution be better when zoomed in? @zeusongit If you dont mind, would you make a comparison using one of the node I updated previously v.s. the one updated by you, because I used 100 ones

File: src/DynamoCoreWpf/Views/Debug/UpdateNodeIconsWindow.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+            {
+              if (value != newIconPaths)
+              {
+                 newIconPaths = value;
+                 OnPropertyChanged(nameof(NewIconPaths));
+              }
+            }
+        }
+        private string output;
+        public string Output
```
Comment (reddyashish) [inline src/DynamoCoreWpf/Views/Debug/UpdateNodeIconsWindow.xaml.cs:49]: Add comments to public properties.

File: src/DynamoCoreWpf/Views/Debug/UpdateNodeIconsWindow.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+            {
+                outputFile.WriteLine(log);
+            }
+        }
+
+        private NodeSearchElement FindNodeFromIconImage(string imgName)
+        {
+            var node = entries.FirstOrDefault(e => e.FullName.ToLower().Equals(imgName));
+            imgName = string.Join(".", imgName.Split('.').Skip(1).ToArray());// because icon name does not include assembly name
+            node = node ?? entries.FirstOrDefault(e => e.IconName.ToLower().Equals(imgName));
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Debug/UpdateNodeIconsWindow.xaml.cs:268]: any other edge cases you saw?

**Comment by zeusongit** [comment]
> > > Nice automation, just curious did you use the 100 subfolder or the 400 subfolder?
> > 
> > 
> > 400 for all
> 
> @Jingyi-Wen I think we were talking about integrating 100 ones but curious if 400 ones would bring any benefit? e.g. would resolution be better when zoomed in? @zeusongit If you dont mind, would you make a comparison using one of the node I updated previously v.s. the one updated by you, because I used 100 ones

In comparison, the 400 icons do look sharper than the 100 ones.
(The top ones are old and bottom is new)
cc

File: src/DynamoUtilities/DebugModes.cs [hunk 3]
```diff
@@ -55,6 +55,7 @@ private static void RegisterDebugModes()
             // Register app wide new debug modes here.
             AddDebugMode("DumpByteCode", "Dumps bytecode to a log file in a folder called ByteCodeLogs located in the current working dirrectory.", false);
             AddDebugMode("CrashOnNewNodeModel", "Crash when creating a new NodeModel. Works only on Debug builds", false);
+            AddDebugMode("UpdateNodeIcons", "This debug mode enables an automation that can be used to update node icons. https://wiki.autodesk.com/x/OACkf", true);
```
[the remark this answers is not in the document]
Comment (zeusongit) [reply src/DynamoUtilities/DebugModes.cs]: I started with this being hidden behind the debug mode, removed now

**Review by QilongTang — APPROVED** [review]

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 4]
```diff
@@ -3370,6 +3372,35 @@ public object ConvertBack(object value, Type targetType, object parameter, Syste
...
+            {
+                s = value as string;
+            }
+
+            if (string.IsNullOrEmpty(s)) return null;
+
+            BitmapImage bi = new BitmapImage();
+
+            bi.BeginInit();
+            bi.StreamSource = new MemoryStream(System.Convert.FromBase64String(s));
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/UI/Converters.cs:3392]: @zeusongit I think you need to carefully examine if this code leaks. You never dispose this stream.

File: src/DynamoCoreWpf/Views/Debug/UpdateNodeIconsWindow.xaml.cs [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+using System.Threading.Tasks;
+using System.Windows;
+using System.Xml.Linq;
+using Dynamo.Search.SearchElements;
+using Dynamo.Utilities;
+
+namespace Dynamo.Wpf.Views.Debug
+{
+    /// <summary>
+    /// Interaction logic for DebugModesWindow.xaml
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Debug/UpdateNodeIconsWindow.xaml.cs:17]: comment incorrect.

**Comment by mjkkirschner** [comment]
@zeusongit this PR brings in lots of new public methods that need to be added to the unshipped.txt file.

**Pull request merged**